# -*- mode: Snakemake -*-
#
# list the all input files for each step

####################
## classify
####################

### Classify reads
TARGET_METAPHLAN = expand(str(CLASSIFY_FP/'metaphlan'/'{sample}.txt'), sample = Samples.keys())
TARGET_MATAPHLAN_REPORT = [str(CLASSIFY_FP/'taxonomic_assignments.tsv')]
TARGET_CLASSIFY = TARGET_METAPHLAN + TARGET_MATAPHLAN_REPORT


####################
## mapping
####################

## KEGG pathway finder
TARGET_PATH = expand(str(MAPPING_FP/'summary'/'pathfinder'/'{sample}_summary.json'), 
                    sample=Samples.keys())
TARGET_PATHREPORT = [str(MAPPING_FP/'ko_assignments.tsv')]

## BILE ACID BLAST
TARGET_BILE = expand(str(MAPPING_FP/'blastn'/'bile_acid'/'{gene}'/'{sample}_R1.blastn'), 
                      gene=GENES_DICT.keys(), sample=Samples.keys())

## FUGAL BOWTIE MAPPING
TARGET_FUNGI = expand(str(MAPPING_FP/'bowtie'/'{genome}'/'summary'/'{sample}_{suffix}'),
                      sample=Samples.keys(), genome=GENOMES_DICT.keys(), 
                      suffix=['counts', 'coverage'])

## ANTIBIOTICS RESISTANCE 
TARGET_ABX = (
    expand(
           str(ANNOTATION_FP/'blastx'/'card'/'{sample}.blastx'),
           sample=Samples.keys()) +
    expand(
           str(MAPPING_FP/'blastx'/'card'/'{sample}_R1.blastx'),
           sample=Samples.keys()))

## GLYCOSIDE HYDROLASE
TARGET_GLYCO = expand(str(MAPPING_FP/'diamond'/'{db}'/'{sample}_R1.m8'),
                      db=DIAMOND_DICT.keys(), sample=Samples.keys())

## MAP READS TO CONTIGS
TARGET_CONTIGS = expand(str(MAPPING_FP/'contigs'/'{sample}_aln.pdf'), 
                        sample=Samples.keys())

TARGET_MAPPING = TARGET_PATH + TARGET_PATHREPORT + TARGET_BILE + TARGET_FUNGI + TARGET_ABX


## TARGETS for sunbeam_neutrino
TARGET_ALL = TARGET_CLASSIFY + TARGET_PATH + TARGET_PATHREPORT
